File: jar-executor-plugin/src/main/java/org/restler/StartProcessMojo.java
package org.restler;

import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import java.util.function.Consumer;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;

import java.io.*;

@Mojo(name = "start", defaultPhase = LifecyclePhase.PRE_INTEGRATION_TEST)
public class StartProcessMojo extends AbstractMojo  {

    @Parameter(required = true)
    private String jar;

    @Parameter(defaultValue = "0")
    private Long delay;

    @Override
    protected void finalize() throws Throwable {
        CurrentProcess.killProcess();
        super.finalize();
    }

    @Override
    public void execute() throws MojoExecutionException, MojoFailureException {
        try {


            if(!jar.endsWith(".jar")) {
                getLog().error("File must be jar.");
                throw new MojoFailureException("File must be jar.");
            }
            else if(!(new File(jar)).exists()) {
                getLog().error("Can't find file " + jar);
                throw new MojoFailureException("Can't find file " + jar);
            }

            Process newProcess = Runtime.getRuntime().exec("java -jar " + jar);
            if(newProcess.isAlive()) {
                getLog().info("Process is started.");
            }

            runInputStreamReaderThread(newProcess.getInputStream(), (String line)->getLog().info(line));
            runInputStreamReaderThread(newProcess.getErrorStream(), (String line)->getLog().error(line));

            if(delay > 0) {
                getLog().info("Start waiting " + delay + "ms ...");
                Thread.sleep(delay);
                if(newProcess.isAlive()) {
                    getLog().info("Stop waiting.");
                }
            }

            if(!newProcess.isAlive()) {
                getLog().info("Process is stopped.");
            }

            CurrentProcess.setProcess(newProcess);
        } catch (IOException | InterruptedException e) {
            getLog().error(e.getMessage());
        }
    }

    private void runInputStreamReaderThread(InputStream inputStream, Consumer<String> printMessage) {
        Thread inputThread = new Thread(() -> {
            try (BufferedReader br = new BufferedReader(new InputStreamReader(inputStream))) {
                String line = br.readLine();
                while (line != null) {
                    printMessage.accept(line);
                    line = br.readLine();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
        inputThread.setDaemon(true);
        inputThread.start();
    }
}

File: jar-executor-plugin/src/main/kotlin/org/restler/StartProcessMojo.kt
package org.restler

import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.plugin.MojoFailureException
import org.apache.maven.plugins.annotations.LifecyclePhase
import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import java.io.*
import java.util.concurrent.CompletableFuture
import java.util.regex.Pattern
import kotlin.concurrent.thread

@Mojo(name = "start", defaultPhase = LifecyclePhase.PRE_INTEGRATION_TEST)
class StartProcessMojo : AbstractMojo() {

    @Parameter(required = true, defaultValue = "")
    private lateinit var jar: String

    @Parameter(defaultValue = "")
    private var waitingLine: String = ""

    @Parameter(defaultValue = "0")
    private val delay: Long = 0

    private var waitingFutureLine: CompletableFuture<Void> = CompletableFuture()
    private lateinit var waitingLinePattern: Pattern

    override fun execute() {
        try {
            if (!jar.endsWith(".jar")) {
                log.error("File must be jar.")
                throw MojoFailureException("File must be jar.")
            } else if (!File(jar).exists()) {
                log.error("Can't find file " + jar)
                throw MojoFailureException("Can't find file " + jar)
            }

            val newProcess = Runtime.getRuntime().exec("java -jar $jar")
            if (newProcess.isAlive) {
                log.info("Process is started.")
            }

            CurrentProcess.setProcess(newProcess);

            waitingLinePattern = Pattern.compile(waitingLine)

            runInputStreamReader(newProcess.inputStream) { checkWaitingLine(it); log.info(it) }
            runInputStreamReader(newProcess.errorStream) { log.error(it) }

            if (delay > 0) {
                log.info("Start waiting " + delay + "ms ...")
                Thread.sleep(delay)
                if (newProcess.isAlive) {
                    log.info("Stop waiting.")
                }
            }

            if(!waitingLine.isEmpty()) {
                waitingFutureLine.get()
            }

            if (!newProcess.isAlive) {
                log.info("Process is stopped.")
            }
        } catch (e: IOException) {
            log.error(e.message)
        } catch (e: InterruptedException) {
            log.error(e.message)
        }
    }

    private fun checkWaitingLine(line: String) {
        if(waitingLinePattern.matcher(line).matches()) {
            waitingFutureLine.complete(null)
        }
    }

    private fun runInputStreamReader(inputStream: InputStream, printMessage: (String) -> Unit) {
        thread(isDaemon = true) {
            BufferedReader(InputStreamReader(inputStream)).use { br ->
                br.forEachLine(printMessage)
            }
        }
    }
}
